### ISSUE TITLE
Specifying 'spec' parameter requires 'url' parameter to work

### ISSUE THREAD

**Issue by ghost** [swagger-api/swagger-ui#1459]
When you attempt to create a new `SwaggerUI` based on a JSON object containing a spec (Swagger 2.0), Swagger fails to render the spec.

The code looks like this (no `url` is specified per the documentation):

```
window.swaggerUi = new SwaggerUi({
    spec: swaggerJsonObj,
    ...
})
```

It fails while trying to access the `url` parameter, which does not exist when `spec` is set to a JSON object.

The failure is happening in `swagger-ui.js`:

```
SwaggerUi.prototype.load = function() {
    var url, _ref1;
    if ((_ref1 = this.mainView) != null) {
        _ref1.clear();
    }
    url = this.options.url;
    if (url.indexOf("http") !== 0) {
        url = this.buildUrl(window.location.href.toString(), url);
    }
    ...
}
```

**Comment by webron** [comment]
Well, the title is inaccurate. It works fine as long as there's also a url parameter.

**Comment by jhaoheng** [comment]
hi webron

Do you mean ?

```
        url: null,
        spec: "swagger.json",
        validatorUrl: null,
```

**Comment by webron** [comment]
@jhaoheng - the `spec` parameter doesn't accept a file name but the actual JSON describing your Swagger.

**Comment by jhaoheng** [comment]
woo
thank you @webron

**Comment by andylowry** [comment]
Followup question: If I provide a spec, how to I prevent the Swagger-UI from attempting to load anything from the provided URL? The README specifically mentions this in the documentation of the `spec` property:

> A JSON object describing the OpenAPI Specification. When used, the url parameter will not be parsed. This is useful for testing manually-generated specifications without hosting them. Works for Swagger 2.0 specs only.

When I provide a spec without a url I'm nothing in the browser, and browser dev tools tell report 

> index.js:24 Uncaught TypeError: Cannot read property 'indexOf' of undefined

If I provide an empty url I see my API but there's a red "Error" marker at the bottom right telling me that my url could not be loaded.

How can I provide a spec and prevent any attempt to load a url?

**Comment by fehguy** [comment]
You still need to provide a url if the spec is provided, but we should be able to support the absence of one by reading the spec itself.  It's just not implemented yet, though.

**Comment by andylowry** [comment]
Thanks. More important to me than being able to leave off the URL is that, if a spec is provided, the URL, if also provided, should be ignored, as promised in the documentation. The scenario suggested there - of a spec that is not hosted anywhere - is exactly my use case, and with v2.1.4 I'm currently at a loss as to how to make it work again.

**Comment by CNoetzel** [comment]
Are there any futher news to this issue? I'm currently struggeling with the same problem.

**Comment by webron** [comment]
You're struggling with adding a url parameter?

**Comment by CNoetzel** [comment]
No, the problem is - as described by andylowry - that the SwaggerUi doesn't behave as mentioned in the documentation. If I provide the json via spec parameter the url parameter is still required, otherwise an error is thrown. Buf if I provide an url and the spec parameter, the json from spec is ignored.

**Comment by webron** [comment]
Are you saying that if you add a `url` and a `spec` parameters, the `spec` is being ignored?

**Comment by andylowry** [comment]
I apologize for not having reported back the way I discovered to make this work propertly, but here it is:

* Supply a model via `spec` parameter
* Supply a bogus URL via the `url` parameter
* Set the `validatorUrl` parameter to `null`

Without the final step, the supplied URL will not be ignored, as claimed in the documentation. So I'd say the README document should be corrected in this respect. Or, preferably, the code should be changed so that if `spec` is provided, `URL` is ignored, as claimed. It would make sense in this case to validate the provided `spec`, and setting `validatorUrl` to `null` prevents this.

**Comment by webron** [comment]
Do you encounter that in the latest version of the ui though? You mentioned v2.1.4 which you really shouldn't be using anymore :)

**Comment by CNoetzel** [comment]
Thanks for the hint, I'll try it tomorrow.

**Comment by andylowry** [comment]
@webron I hadn't tried it since I found a work-around. I have now, and can confirm that in current master branch, with a `spec` provided, the `url` value appears to be ignored.

I would not object to closing this issue - can't speak for the others who have posted to the issue (including the originator).

**Comment by fehguy** [comment]
This is a bug.  The `url` parameter needs to be explicitly specified, and the one in the spec itself is ignored.

**Comment by webron** [comment]
@fehguy any chance you confused `url` with the ability to override the `host`?

**Comment by fehguy** [comment]
well, from the title of this issue, the problem is that if the `url` is not specified, the UI will fail to load.  The spec object itself has everything necessary to compute that URL, and therefore it should probably just honor it when a model is passed rather than the `url` alone.

Thus, I think the fix is to follow this logic:

* If the URL is supplied and spec is not, use URL to load the spec.
* If the URL is not supplied and the spec is, compute the URL from the spec and set it in the UI
* If the URL is specified _and_ the spec is specified, use the URL value

Does that make sense?  Apologies if I made things more confusing.

**Comment by CNoetzel** [comment]
I thought the spec is a json object, containing the complete swagger definition. Do you really need to compute the URL from the spec? Isn't the URL in that case (spec provided) needless?

**Comment by fehguy** [comment]
Yes you do need to compute it, since host, schemes, basePath are all optional.  But that is the bug.

**Comment by andylowry** [comment]
@fehguy I honestly think you've got things confused, apologies if I'm wrong.

The `url` parameter is used by swagger-ui to load the spec. If the `spec` parameter is provided, no `url` is needed to load the spec, since the spec is supplied.

This has nothing to do with `host`, `schemes`, and `basePath` within the spec. Those do not address the `spec` - they address the modeled service endpoint, which is displayed in the rendered HTML and is also used in the "Try it out" feature.

The URL we're talking about is _not_ rendered anywhere in the HTML and is not used in "Try it out." It is only used to load the spec.

At least that's my understanding.

**Comment by fehguy** [comment]
I don't know why we're arguing over misunderstandings, but as I said, this is a known bug.

**Comment by andylowry** [comment]
@fehguy Let me try state my position one more time, then if you still disagree I'll let it go.

The bug that gave rise to this issue _not_ a bug anymore, known or otherwise. The original bug had to do with how one could supply a spec to swagger-ui via `url` and `spec` parameters to the `SwaggerUI` constructor. It was not working the way the README file said it should. Now it does. That bug is fixed.

You mention `host` and `schemes` and `baseUrl` properties within the spec, which have nothing to do with how to supply the spec to swagger-ui. This is why I'm convinced you're confusing this issue with something else. There probably are some known bugs that have to do with `host`, `schemes` and `baseUrl`, but they're not what this issue is about. In my opinion, this issue should be closed.

My reasons for continuing this discussion are (a) to prevent this discussion leaving future readers confused; and (b) more importantly, to allow this issue to be closed. If it did represent an active, known bug, as you have stated, that should prevent it being closed.

**Comment by fehguy** [comment]
@andylowry I don't get it.  Read the title of the issue, that's what I'm talking about.  If you have something else, open a new issue or you will confuse people.

For everyone else who stumbles upon here, there *is a bug* in swagger-ui, which _requires_ the `url` to be specified even when supplying the spec *AND* the spec has enough information to deduce the URL, and that's why this bug is still open.

**Comment by andylowry** [comment]
@fehguy Sorry, you're right - `null` is still not allowed in `url` when `spec` is supplied. A bogus URL now causes no problem, and that solves the situation that originally caused me to chime in on the issue.

But what I still just don't get is what you're talking about when you mention `host` and `baseUrl` and `schemes`, and "the spec [having] enough information to deduce the URL." Now that I understand that you really are talking about the URL one would use to retrieve the spec, I'm left with this question: How could those or any other values appearing in the spec allow one to deduce a URL for obtaining the spec? Those properties are about a service endpoint for the API described by the spec, not the location of the spec itself. Is there some convention by which those are assumed to be related?

**Comment by andylowry** [comment]
@fehguy Looking back through the discussion, I can see that even as far back as my initial comment over a year ago, you responded with something to the effect that the URL could be computed from the spec. 

In the recent back-and-forth when I saw similar comments I thought you were confusing issues, because it never occurred to me that you might actually mean that. They were apples and oranges to my mind.

This concept still leaves me baffled. Would really appreciate a hint as to how one might deduce the spec's URL from the spec.

**Comment by tedepstein** [comment]
@fehguy wrote:

> Thus, I think the fix is to follow this logic:
>
> - If the URL is supplied and spec is not, use URL to load the spec.
> - If the URL is not supplied and the spec is, compute the URL from the spec and set it in the UI
> - If the URL is specified and the spec is specified, use the URL value

If OpenAPI v3 gives us an opportunity to take a fresh look at this, I'd suggest an alternate way to fix the logic:

* Rename the `url` parameter to `specURL` to clarify its intended usage as location from which Swagger-UI should load the OpenAPI spec.  ("Rename" could just mean "deprecate the old name, with some kind of visible warning.")
* Rename the `spec` parameter to `specObject` for consistency & clarity.
* Make `specURL` and `specObject` arguments mutually exclusive.  Exactly one of these args should be be provided; the other must be omitted or must have a `null` value. 
* Introduce a new `serviceURL` argument that acts as an override to the `host`, `schemes` and `basePath` included in the spec.  This would work the same, regardless of whether the spec was passed via `specURL` or `specObject`, and it helps address scenarios like the one we talked about [here](https://github.com/swagger-api/swagger-codegen/issues/4529), where OpenAPI spec is created a design time, but the physical endpoint is determined at runtime.  I think this override behavior is what you meant in your third bullet point. 

Would this work?

**Comment by webron** [comment]
This is solved in 3.0.
